docs/design/object-locking.md

```diff
@@ -0,0 +1,207 @@
+`s3:PutObjectLegalHold` are allowed to set Legal hold on an object version.
+
+### Implementation
+
+#### Storing Object lock configuration
+
+Object lock can be enabled on a bucket during bucket creation using
+`x-amz-bucket-object-lock-enabled` header. This is stored along with the
+bucket's metadata defined in the BucketInfo model.
+Any default lock settings (governance/compliance mode, days/years for retention
+period) defined using `PutObjectLockConfiguration` request are stored
+as part of the bucket metadata.
+When a PUT Object request is received, the lock configuration is evaluated and
+the lock expiration date and time is calculated and set as `retain-until-date`
+property on the object's metadata.
+Lock configuration on a PUT object request is evaluated the following order
+
+1. Bucket's configuration for retention mode/period
+2. Object's configuration in the PUT Object for retention mode/period
+
+The object's settings override the bucket's settings when calculating the
+`retain-until-date` date and time to be stored on the object's metadata.
+The lock configuration on an object version can also be changed using
+`PutObjectRetention` api request, when the object has no prior retention mode
+set or if the object has `GOVERNANCE` mode set and the user has appropriate
+permission to make the request. When applying `PutObjectRetention` configuration
+on an object that already has a `retain-until-date` set, the newer configuration
+can only be greater than or equal to the current `retain-until-date`, it cannot
+be shortened.
+
+Note: Delete markers do not have any object lock protection
+
+#### Processing DELETE requests
+
+Whenever requests such as DELETE object using version-id,
+Multi-Object DELETE specifying version-id or a Lifecycle action to
+permanently delete the object version are received, the current date and time is
+evaluated against the `retain-until-date` set on the object and the client will
+receive an `Access Denied` error if the current date and time is less than the
+`retain-until-date` set on the object.
+When the current date and time exceeds the `retain-until-date`, deletes on the
+object version are automatically allowed. There is no cleanup action to remove
+the `retain-until-date` set on the object version's metadata once the
+`retain-until-date` expires.
+DELETE object requests without a version id result in creation of delete markers
+on top of the object version, even if the object has a lock configuration set
+and the `retain-until-date` is current.
+
+#### Lifecycle actions
+
+Lifecycle jobs can create delete markers on the object but cannot permanently
+delete an object version that is locked until the lock expires.
+
+#### Replication
+
+The lock configuration on the object version in a source bucket is copied over
+to the destination bucket only if object lock is enabled on the destination
+bucket. Otherwise, the lock is ignored in the destination bucket.
+
+## APIs covering S3 Object lock
```

Comment:
Can we add `Get Object` as well? 

When we have object lock configuration on an object, AWS Get Object request returns ObjectLockMode and ObjectLockRetainUntilDate properties; sample response: 
```
{
    "AcceptRanges": "bytes",
    "LastModified": "2020-05-14T03:59:59+00:00",
    "ContentLength": 2248,
    "ETag": "\"03dfecec5e65f5cacc56bfe6a6fd07d6\"",
    "VersionId": "f0lAvcTBiDKHgqZDpdkAlAoDFG9S0N38",
    "ContentType": "binary/octet-stream",
    "Metadata": {},
    "ObjectLockMode": "GOVERNANCE",
    "ObjectLockRetainUntilDate": "2020-05-15T03:59:58.451000+00:00"
}
```